### PR TITLE
HTTPCLIENT-2185 Add support for IDNA 2008 (RFC 5891)

### PR DESCRIPTION
**Pull request by hanasakari**
support IDNA 2008 (RFC 5896)  for [Add support for IDNA 2008 (RFC 5891)](https://issues.apache.org/jira/browse/HTTPCLIENT-2185?jql=project%20in%20(HTTPCLIENT%2C%20HTTPCORE)%20AND%20status%20%3D%20Open%20AND%20labels%20%3D%20volunteers-wanted%20AND%20assignee%20in%20(EMPTY))

### CODE REVIEW

File: httpcore5/src/main/java/org/apache/hc/core5/net/Host.java [hunk 1]
```diff
@@ -71,7 +73,10 @@ static Host parse(final CharSequence s, final Tokenizer.Cursor cursor) throws UR
             if (!InetAddressUtils.isIPv6Address(hostName)) {
                 throw URISupport.createException(s, cursor, "Expected an IPv6 address");
             }
-        } else {
+        }  else if (!StandardCharsets.US_ASCII.newEncoder().canEncode(s.toString())){
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/net/Host.java:76]: @hanasakari I suspect this operation can be relative expensive. Could you please add a method to `TextUtils` that performs the same test without extra object allocation? 
Please also note that it is likely to be more efficient to execute the test and IDN transformation _after_ `tokenizer#parseContent` parsing operation.

**Review by ok2c — CHANGES_REQUESTED** [review]

**Comment by garydgregory** [comment]
Is it worth caching the encoder?

**Comment by ok2c** [comment]
> Is it worth caching the encoder?

No, it is not. The idea is to not use an encoder at all.

**Comment by ok2c** [comment]
@hanasakari That is exactly what I wanted. Cool.Almost there.
1. The build currently fails due to trailing blanks somewhere. That needs to be fixed.
2. I suppose this bit could be optimized
```
        } else if (!TextUtils.isAllASCII(s)) {
            hostName = IDN.toASCII(tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS));
        }
        else {
            hostName = tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS);
        }
```        
to
```
        else {
            hostName = tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS);
            if (!TextUtils.isAllASCII(hostName)) {
                hostname = IDN.toASCII(hostname)
            }
        }
```        
3. #format method needs to be symmetrically adjusted.

Oleg

**Comment by hanasakari** [comment]
ok，i fix it

**Comment by ok2c** [comment]
@hanasakari What about other changes?

**Comment by hanasakari** [comment]
this code

```
        else {
            hostName = tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS);
            if (!TextUtils.isAllASCII(hostName)) {
                hostname = IDN.toASCII(hostname)
            }
        }
```
IDE tips
Variable 'hostName' might already have been assigned to

so change to

```
// If the hostName contains non-ASCII characters, IDNA processing is required
hostName = TextUtils.isAllASCII(s) ?
      tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS) :
      IDN.toASCII(tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS));
```

**Comment by ok2c** [comment]
@hanasakari There is still a problem.  The routine must work with the content of `s within the bounds specified by the `cursor`. Your code does not take the `cursor` into account when executing `TextUtils#isAllASCII`. This is the reason the content of `s` should be parsed first and then tested with `TextUtils#isAllASCII`. Make `hostName` non-final if needed.

File: httpcore5/src/main/java/org/apache/hc/core5/net/Host.java [hunk 2]
```diff
@@ -72,7 +73,10 @@ static Host parse(final CharSequence s, final Tokenizer.Cursor cursor) throws UR
                 throw URISupport.createException(s, cursor, "Expected an IPv6 address");
             }
         } else {
-            hostName = tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS);
+            // If the hostName contains non-ASCII characters, IDNA processing is required
+            hostName = TextUtils.isAllASCII(s) ?
+                    tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS) :
+                    IDN.toASCII(tokenizer.parseContent(s, cursor, URISupport.PORT_SEPARATORS));
```
Comment (j-bernard) [inline httpcore5/src/main/java/org/apache/hc/core5/net/Host.java:79]: java.net.IDN is implementing a deprecated version of IDNA (IDNA 2003). The latest version, IDNA 2008, can be supported using the ICU4j library

**Pull request closed without merging**